crates/ruff_python_semantic/src/model.rs

```diff
@@ -836,22 +839,32 @@ impl<'a> SemanticModel<'a> {
 
     /// Return the current `Expr`.
     pub fn expr(&self) -> Option<&'a Expr> {
-        self.exprs.iter().last().copied()
+        let node_id = self.expr_id?;
+        Some(self.exprs[node_id])
     }
 
-    /// Return the parent `Expr` of the current `Expr`.
+    /// Return the parent `Expr` of the current `Expr`, if any.
     pub fn expr_parent(&self) -> Option<&'a Expr> {
-        self.exprs.iter().rev().nth(1).copied()
+        self.expr_ancestors().next()
     }
 
-    /// Return the grandparent `Expr` of the current `Expr`.
+    /// Return the grandparent `Expr` of the current `Expr`, if any.
     pub fn expr_grandparent(&self) -> Option<&'a Expr> {
-        self.exprs.iter().rev().nth(2).copied()
+        self.expr_ancestors().nth(1)
     }
 
     /// Return an [`Iterator`] over the current `Expr` parents.
-    pub fn expr_ancestors(&self) -> impl Iterator<Item = &&Expr> {
-        self.exprs.iter().rev().skip(1)
+    pub fn expr_ancestors(&self) -> impl Iterator<Item = &'a Expr> + '_ {
+        self.expr_id
+            .iter()
+            .copied()
+            .flat_map(|id| {
+                self.exprs
+                    .ancestor_ids(id)
+                    .skip(1)
+                    .map(|id| &self.exprs[id])
+            })
+            .copied()
```

Comment:
Nit: Use successors? 

```suggestion
std::iter::successors(self.expr_id, |id| self.exprs.ancestor_ids(id).skip(1).map(|id| self.exprs[id])))
```

Reply:
This doesn't quite do the same thing. The awkwardness is that `self.expr_id` can be `None`.

Reply:
I'm not sure if I understand. The first argument to `std::iter::successors` is an `Option`. Both `successors and `expr_id.flat_map` should short circuit if expr_id is `None`.

Reply:
Oh, I think you just had misplaced parentheses in your suggestion and I misunderstood.